x/bsonx/bsoncore/array_test.go

```diff
@@ -0,0 +1,306 @@
+			_, got := rdr.IndexErr(1)
+			want := NewInsufficientBytesError(nil, nil)
+			if !compareErrors(got, want) {
+				t.Errorf("Did not receive expected error. got %v; want %v", got, want)
+			}
+		})
+		testCases := []struct {
+			name  string
+			rdr   Array
+			index uint
+			want  Element
+		}{
+			{"first",
+				Array{0xe, 0x0, 0x0, 0x0, 0xa, 0x78, 0x0, 0xa, 0x79, 0x0, 0xa, 0x7a, 0x0, 0x0},
+				0, Element{0x0a, 0x78, 0x00},
+			},
+			{"second",
+				Array{0xe, 0x0, 0x0, 0x0, 0xa, 0x78, 0x0, 0xa, 0x79, 0x0, 0xa, 0x7a, 0x0, 0x0},
+				1, Element{0x0a, 0x79, 0x00},
+			},
+			{"third",
+				Array{0xe, 0x0, 0x0, 0x0, 0xa, 0x78, 0x0, 0xa, 0x79, 0x0, 0xa, 0x7a, 0x0, 0x0},
+				2, Element{0x0a, 0x7a, 0x00},
+			},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				t.Run("IndexErr", func(t *testing.T) {
+					got, err := tc.rdr.IndexErr(tc.index)
+					if err != nil {
+						t.Errorf("Unexpected error from IndexErr: %s", err)
+					}
+					if diff := cmp.Diff(got, tc.want); diff != "" {
+						t.Errorf("Arrays differ: (-got +want)\n%s", diff)
+					}
+				})
+				t.Run("Index", func(t *testing.T) {
+					defer func() {
+						if err := recover(); err != nil {
+							t.Errorf("Unexpected error: %v", err)
+						}
+					}()
+					got := tc.rdr.Index(tc.index)
+					if diff := cmp.Diff(got, tc.want); diff != "" {
+						t.Errorf("Arrays differ: (-got +want)\n%s", diff)
+					}
+				})
+			})
+		}
+	})
+	t.Run("NewArrayFromReader", func(t *testing.T) {
+		testCases := []struct {
+			name     string
+			ioReader io.Reader
+			arr      Array
+			err      error
+		}{
+			{
+				"nil reader",
+				nil,
+				nil,
+				ErrNilReader,
+			},
+			{
+				"premature end of reader",
+				bytes.NewBuffer([]byte{}),
+				nil,
+				io.EOF,
+			},
+			{
+				"empty Array",
+				bytes.NewBuffer([]byte{5, 0, 0, 0, 0}),
+				[]byte{5, 0, 0, 0, 0},
+				nil,
+			},
+			{
+				"non-empty Array",
+				bytes.NewBuffer([]byte{
+					'\x1B', '\x00', '\x00', '\x00',
+					'\x02',
+					'0', '\x00',
+					'\x04', '\x00', '\x00', '\x00',
+					'\x62', '\x61', '\x72', '\x00',
+					'\x02',
+					'1', '\x00',
+					'\x04', '\x00', '\x00', '\x00',
+					'\x62', '\x61', '\x7a', '\x00',
+					'\x00',
+				}),
+				[]byte{
+					'\x1B', '\x00', '\x00', '\x00',
+					'\x02',
+					'0', '\x00',
+					'\x04', '\x00', '\x00', '\x00',
+					'\x62', '\x61', '\x72', '\x00',
+					'\x02',
+					'1', '\x00',
+					'\x04', '\x00', '\x00', '\x00',
+					'\x62', '\x61', '\x7a', '\x00',
+					'\x00',
+				},
+				nil,
+			},
+		}
+
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				arr, err := NewArrayFromReader(tc.ioReader)
+				if !compareErrors(err, tc.err) {
+					t.Errorf("errors do not match. got %v; want %v", err, tc.err)
+				}
+				if !bytes.Equal(tc.arr, arr) {
+					t.Errorf("Arrays differ. got %v; want %v", tc.arr, arr)
+				}
+			})
+		}
+	})
+	t.Run("DebugString", func(t *testing.T) {
```

Comment:
Added some tests for `bsoncore.Array#String()` and `bsoncoreArray#DebugString()`.